mhs-reference-implementation/mhs/handler/async_response_handler.py

```diff
@@ -1,46 +1,59 @@
 import logging
+from typing import Callable, Dict
 
-from tornado.web import RequestHandler
+from tornado.web import RequestHandler, HTTPError
 
 import mhs.builder.ebxml_ack_message_builder as ack_builder
 import mhs.builder.ebxml_message_builder as builder
 import mhs.parser.ebxml_message_parser as parser
-from mhs.sender.sender import PARTY_ID
 
 
 class AsyncResponseHandler(RequestHandler):
-    """A RequestHandler for asynchronous responses from a remote MHS."""
+    """A RequestHandler for asynchronous responses from a remote MHS. Extracts a reference to a previous message from
+    messages received and calls the callback registered against this ID."""
 
-    def initialize(self, ack_builder, message_parser):
+    def initialize(self, ack_builder: builder.EbXmlMessageBuilder, message_parser: parser.EbXmlMessageParser,
+                   callbacks: Dict[str, Callable[[str], None]], party_id: str):
         """Initialise this request handler with the provided dependencies.
 
         :param ack_builder: The message builder to use when building ebXML acknowledgement messages.
         :param message_parser: The message parser to use to parse ebXML asynchronous responses.
-        :return:
+        :param callbacks: The dictionary of callbacks to use when a message is received.
+        :param party_id: The party ID of this MHS. Sent in ebXML acknowledgements.
         """
         self.ack_builder = ack_builder
         self.message_parser = message_parser
+        self.callbacks = callbacks
+        self.party_id = party_id
```

Comment:
I don't quite understand the purpose of the `party_id` but do you feel that this is something that is required to construct the `async_response_handler` or is this something that could be passed in via the `post` method below?

Reply:
The party ID is the identifier of the MHS instance, so it's something you're assigned when you're registered with Spine/Opentest/etc. In this case I thought it made more sense to have the response handler set up with the value than take it from the "to" field of the incoming messages.